url73e3057d61_empeeoh_BACnet4J_tgz-pJ8-com_serotonin_bacnet4j_test_OverflowTestJ8/src/com/serotonin/bacnet4j/service/confirmed/AtomicReadFileRequest.java

```
@@ -0,0 +1,198 @@
+            }
+
+            // Check for status (backup/restore)
+            BackupState bsOld = (BackupState) localDevice.getConfiguration().getProperty(
+                    PropertyIdentifier.backupAndRestoreState);
+            if (bsOld.intValue() == BackupState.preparingForBackup.intValue()
+                    || bsOld.intValue() == BackupState.preparingForRestore.intValue())
+                // Send error: device configuration in progress as response
+                throw new BACnetServiceException(ErrorClass.device, ErrorCode.configurationInProgress);
+
+            file = (FileObject) obj;
+
+            // Validation.
+            FileAccessMethod fileAccessMethod = (FileAccessMethod) file
+                    .getProperty(PropertyIdentifier.fileAccessMethod);
+            if (recordAccess && fileAccessMethod.equals(FileAccessMethod.streamAccess) || !recordAccess
+                    && fileAccessMethod.equals(FileAccessMethod.recordAccess))
+                throw new BACnetErrorException(getChoiceId(), ErrorClass.object, ErrorCode.invalidFileAccessMethod);
+        }
+        catch (BACnetServiceException e) {
+            throw new BACnetErrorException(getChoiceId(), e);
+        }
+
+        if (recordAccess)
+            throw new NotImplementedException();
+
+        long start = fileStartPosition.longValue();
+        long length = requestedCount.longValue();
+
+        /*
+         * throw an exception when the following conditions are met - start is a negative number - start exceeds the
+         * length of the file object
+         */
+        if (start < 0 || start > file.length())
+            throw new BACnetErrorException(getChoiceId(), ErrorClass.object, ErrorCode.invalidFileStartPosition);
+
+        try {
+            response = new AtomicReadFileAck(new Boolean(file.length() <= start + length), fileStartPosition,
+                    file.readData(start, length));
+        }
+        catch (IOException e) {
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)